library/alloc/src/slice.rs

```diff
@@ -1036,149 +1040,150 @@ where
+    // Unchecked array access gives an approximate 20% performance improvement.
+    //
+    // # Safety
+    //
+    // `$left` and `$right` must be < `$v.len()`
+    macro_rules! gt {
+        ($v: ident, $left: expr, $right: expr, $is_less: ident) => {
+            // $is_less(&$v[$right], &$v[$left])
+            {
+                debug_assert!($left < $v.len() && $right < $v.len());
+                $is_less(unsafe { &$v.get_unchecked($right) }, unsafe { &$v.get_unchecked($left) })
+            }
+        };
+    }
+
+    // Benchmarking determined these are the best sizes.
+    // Recursive merge switches to insertion sort / merge when slice length is <= SMALL_SLICE_LEN*2.
+    const SMALL_SLICE_LEN: usize = 10;
     // Slices of up to this length get sorted using insertion sort.
     const MAX_INSERTION: usize = 20;
-    // Very short runs are extended using insertion sort to span at least this many elements.
-    const MIN_RUN: usize = 10;
 
     // Sorting has no meaningful behavior on zero-sized types.
     if size_of::<T>() == 0 {
         return;
     }
 
     let len = v.len();
-
     // Short arrays get sorted in-place via insertion sort to avoid allocations.
     if len <= MAX_INSERTION {
-        if len >= 2 {
-            for i in (0..len - 1).rev() {
-                insert_head(&mut v[i..], &mut is_less);
-            }
+        for i in 1..len {
+            insert_end(&mut v[..=i], &mut is_less);
         }
         return;
     }
 
     // Allocate a buffer to use as scratch memory. We keep the length 0 so we can keep in it
     // shallow copies of the contents of `v` without risking the dtors running on copies if
-    // `is_less` panics. When merging two sorted runs, this buffer holds a copy of the shorter run,
-    // which will always have length at most `len / 2`.
-    let mut buf = Vec::with_capacity(len / 2);
-
-    // In order to identify natural runs in `v`, we traverse it backwards. That might seem like a
-    // strange decision, but consider the fact that merges more often go in the opposite direction
-    // (forwards). According to benchmarks, merging forwards is slightly faster than merging
-    // backwards. To conclude, identifying runs by traversing backwards improves performance.
-    let mut runs = vec![];
-    let mut end = len;
-    while end > 0 {
-        // Find the next natural run, and reverse it if it's strictly descending.
-        let mut start = end - 1;
-        if start > 0 {
-            start -= 1;
-            unsafe {
-                if is_less(v.get_unchecked(start + 1), v.get_unchecked(start)) {
-                    while start > 0 && is_less(v.get_unchecked(start), v.get_unchecked(start - 1)) {
-                        start -= 1;
+    // `is_less` panics. When merging two slices, this buffer holds a copy of the right-hand slice,
+    // which will always have length at most `(len + 1) / 2`.
+    let mut buf = Vec::with_capacity((len + 1) / 2);
```

Comment:
Can this become an array rather than a `Vec`?

Reply:
No, because it needs to be allocated with a length based on the length of the vec to be sorted.

Reply:
Using const generics?

Reply:
But the length of the slice to be sorted is not const.

Reply:
I was thinking of doing the capacity computation as a constant expression, but I see [this isn't supported](https://rust-lang.github.io/rfcs/2000-const-generics.html#when-a-const-variable-can-be-used) (yet).

Reply:
If you call `merge_sort` on with a slice of 18446744073709551615 elements, this will overflow.  

```suggestion
    let mut buf = Vec::with_capacity((len / 2) + 1);
```
should fix that.

Reply:
If you call `merge_sort` on with a slice of 18446744073709551615 elements,`len` will overflow.  

```suggestion
    let mut buf = Vec::with_capacity((len / 2) + 1);
```
should fix that.

Reply:
The suggestion would change the semantics... `(10 + 1) / 2` != `(10 / 2) + 1`.  Performance is sensitive to how this is calculated.

> If you call merge_sort on with a slice of 18446744073709551615 elements,len will overflow.

But it is not possible to create a slice that big.

Reply:
This could show up as possible integer overflow using analysis tools though, and why not prevent that.

#88581 hasn't landed yet.
Can you use [`saturating_add`](https://doc.rust-lang.org/std/primitive.usize.html#method.saturating_add) instead?

(Don't know if there is some API to e.g., represent memory a mapped file as an extremely large slice or so, or some extreme supercomputer in the future, but the practicality of such a slice wasn't my point.)

Reply:
`saturating_add` would create a slice that is too short by `1`.  However it is not possible to create a slice that is the size of memory, and then allocate an additional slice that is 1/2 that size.  If an extreme supercomputer existed that could represent this, then the max value of `usize` would also have to increase.

Reply:
To be clear, the change I proposed as fallback should calculate an equal capacity unless overflow occurs:

```suggestion
    let mut buf = Vec::with_capacity(len.saturating_add(1) / 2);
```

If you mean this change would prevent an overflow but leave `buf` possibly one element to small, then you may be right (the wording is that 'at most' this scratch space is needed, if we can prove this isn't so in this corner case, the argument would not hold). 

You would avoid underallocating with my alternate formula, that overallocates one element for even sizes of `self`. Your scratch buffer procedure cannot work at `len == 18446744073709551615` anyway (under 64 bits word size, but mind lower word sizes and memory mapped files), since as you write, the rest of the algorithm requires more than half of that size in space. 

Alternatively, you may simply store one overflow scratch element in a variable, and keep the scratch space at a neat `self.len() / 2` length.

I do not know exactly the guaranteed properties of slices other than being 'contiguous'. But in my mind, a sparse bit vector could be a slice with 18446744073709551615 elements but take far fewer bytes in storage. So the range of `usize` would not have to increase in that case. I just hope to encourage you to think about this, since this is fundamental standard library function for which we cannot make assumptions.

By the way, the (current and original) wording in the API docs that it allocates 'half' the size of `self` in scratch space is subtly incorrect, then.

Reply:
> But it is not possible to create a slice that big.

With zero-sized types it is.

```rust
fn main() {
    [(); usize::MAX];  // works
}
```

Reply:
That's already prevented

```rust
    // Sorting has no meaningful behavior on zero-sized types.
    if size_of::<T>() == 0 {
        return;
    }
```